x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -332,6 +339,146 @@ private void createApiKeyAndIndexIt(
+        }
+        if (Strings.isNullOrEmpty(apiKeyDoc.name)) {
+            throw new ValidationException().addValidationError("cannot update legacy api key [" + apiKeyId + "] without name");
+        }
+        // TODO also assert that authentication subject matches creator on apiKeyDoc
+    }
+
+    private static <T> T single(Collection<T> elements) {
+        if (elements.size() != 1) {
+            throw new IllegalStateException("collection must have exactly one element but had [" + elements.size() + "]");
+        }
+        return elements.iterator().next();
+    }
+
+    private static <T> T single(T[] elements) {
+        if (elements.length != 1) {
+            throw new IllegalStateException("array must contain exactly one element but had [" + elements.length + "]");
+        }
+        return elements[0];
+    }
+
+    private boolean isActive(ApiKeyDoc apiKeyDoc) {
+        return apiKeyDoc.invalidated == false
+            && (apiKeyDoc.expirationTime == -1 || Instant.ofEpochMilli(apiKeyDoc.expirationTime).isAfter(clock.instant()));
+    }
+
+    private ResourceNotFoundException apiKeyNotFound(String apiKeyId) {
+        return new ResourceNotFoundException("api key [" + apiKeyId + "] not found");
+    }
+
+    private void doBulkUpdate(BulkRequest bulkRequest, ActionListener<BulkResponse> listener) {
+        securityIndex.prepareIndexIfNeededThenExecute(
+            listener::onFailure,
+            () -> executeAsyncWithOrigin(client.threadPool().getThreadContext(), SECURITY_ORIGIN, bulkRequest, listener, client::bulk)
+        );
+    }
+
+    private BulkRequest buildBulkUpdateRequest(
+        Authentication authentication,
+        UpdateApiKeyRequest request,
+        Set<RoleDescriptor> userRoles,
+        Collection<ApiKeyDocWithSeqNoAndPrimaryTerm> apiKeyDocs
+    ) throws IOException {
+        final var version = clusterService.state().nodes().getMinNodeVersion();
+        final var bulkRequestBuilder = client.prepareBulk();
+        for (ApiKeyDocWithSeqNoAndPrimaryTerm apiKeyDoc : apiKeyDocs) {
+            bulkRequestBuilder.add(buildIndexRequestForUpdate(authentication, request, userRoles, version, apiKeyDoc));
+        }
+        bulkRequestBuilder.setRefreshPolicy(RefreshPolicy.WAIT_UNTIL);
+        return bulkRequestBuilder.request();
+    }
+
+    private IndexRequest buildIndexRequestForUpdate(
+        Authentication authentication,
+        UpdateApiKeyRequest request,
+        Set<RoleDescriptor> userRoles,
+        Version version,
+        ApiKeyDocWithSeqNoAndPrimaryTerm apiKeyDocWithSeqNoAndPrimaryTerm
+    ) throws IOException {
+        final var apiKeyDoc = apiKeyDocWithSeqNoAndPrimaryTerm.apiKey();
+        final var newRoleDescriptors = request.getRoleDescriptors();
+        // TODO feels backwards that we deserialize just to serialize again
+        final var keyRoles = newRoleDescriptors != null
```

Comment:
This part I'm not sure about:

It seems odd to take the already serialized roles from the doc, deserialize them, then serialize them in `newDocument`. Same for metadata.

Ideally, I would just take the `ByteReference` instances and directly include them in the source but I haven't found an appropriate way to do this yet.

Wondering if you had suggestions, or whether you think the current approach is fine.

Reply:
You can use something like the follows:
```
builder.rawField("role_descriptors", apiKeyDoc.roleDescriptorsBytes, XContentType.JSON);
```